### PR TITLE
Terminal Error handler

### LINKED ISSUE #53: Error handling improvements
When the router catches an exception it enters the Halted state. Any further exception should rethro the initial exception (like esp-net).

The `Router` should be able to take a TerminalErrorHandler, an object with a single method `onError(err)`. If this is provided at construction time errors will be delegated to it rather than thrown.

### PR DESCRIPTION
**Pull request by raybooysen**
When the router has an error purging event queues at any stage, log it and pass it to the error handler.

Fixes #53

### CODE REVIEW

File: src/router/Router.js [hunk 1]
```diff
@@ -279,56 +291,63 @@ export default class Router extends DisposableBase {
...
-                this._state.moveToDispatchModelUpdates();
-                this._dispatchModelUpdates();
-                modelRecord = this._getNextModelRecordWithQueuedEvents();
-                hasEvents = typeof modelRecord !== 'undefined';
-            } // keep looping until any events from the dispatch updates stages are processed
+                } // keep looping until any events from the dispatch updates stages are processed
+            }
+            catch(err) {
+                _log.error('Router saw an error purging queues: [{0}]', err);
+                this._onErrorHandler(err);
```
Comment (KeithWoods) [inline src/router/Router.js]: I think this try catch and dispatch is in the wrong place. There is existing halting flows we should place the dispatch to the error handler there. The `_halt(err)` is the place to put it. If you want to distinguish between the initial exception and follow-on exceptions we can do that within the `_halt()` method. For example any interaction with the router after halting will results in further calls to `_halt()`, that method could query `_state` and determine if it's the initial exception or a follow-on.

File: src/router/Router.js [hunk 2]
```diff
@@ -164,6 +169,13 @@ export default class Router extends DisposableBase {
             return this._observeEventsUsingConventions(modelId, object, methodPrefix);
         }
     }
+    addOnErrorHandler(handler) {
+        this._onErrorHandler = handler;
```
Comment (KeithWoods) [inline src/router/Router.js]: This method name would imply there is a collection of error handlers, however the functionality actually replaces the last or default registered one.

I think the name is correct, however the underlying store should be an array of handlers.

**Comment by KeithWoods** [comment]
Looks good. Few comments above. Other thing is there should be some unit tests covering the functionality.

**Comment by KeithWoods** [comment]
linking issue #53

File: src/router/Router.js (hunk 1 above)
Comment (raybooysen) [inline src/router/Router.js]: Good point.

File: src/router/Router.js (hunk 2 above)
Comment (raybooysen) [inline src/router/Router.js]: sounds good.

File: src/router/Router.js (hunk 1 above)
Comment (raybooysen) [inline src/router/Router.js]: Ah I see.  I didn't see that each called to _purgeEventQueues had a try/catch

**Comment by raybooysen** [comment]
Updated based on your comments.  Please review.

File: src/router/Router.js [hunk 3]
```diff
@@ -164,6 +167,18 @@ export default class Router extends DisposableBase {
...
     }
+    addOnErrorHandler(handler) {
+        this._onErrorHandlers.push(handler);
+    }
+    removeOnErrorHandler(handler) {
+        let index = this._onErrorHandlers.indexOf(handler);
+        if(index >= 0) {
+            delete this._onErrorHandlers[index];
+        } else {
+            throw new Error('Unknown error handler.');
```
Comment (KeithWoods) [inline src/router/Router.js]: should we really throw here?
It's always one of those questions with remove calls, if it's not there then all you were doing was removing it anyway, but on the other hand if you silently carry on it eludes to a bug in caller code. 

So in summery this is fine :)

File: src/router/Router.js [hunk 4]
```diff
@@ -465,11 +480,31 @@ export default class Router extends DisposableBase {
         }
     }
     _halt(err) {
+        //Storing the current status locally
+        // as moveToHalted() may move the state onwards
+        // and we only want to run the error handlers on the first error
+        let localStatus = this._state.currentStatus;
```
Comment (KeithWoods) [inline src/router/Router.js]: can we change this so the `Status.Halted` check happens here, then we give this variable name something more descriptive. 

For example:
```
let isInitialHaltingError = this._state.!== Status.Halted

// then later 

if(isInitialHaltingError) {
}
```
you probably don't need the comments then as it's fairly self descriptive (hopefully :)).

Comment (raybooysen) [inline src/router/Router.js]: Agreed.

File: src/router/Router.js [hunk 5]
```diff
@@ -164,6 +167,18 @@ export default class Router extends DisposableBase {
             return this._observeEventsUsingConventions(modelId, object, methodPrefix);
         }
     }
+    addOnErrorHandler(handler) {
+        this._onErrorHandlers.push(handler);
```
Comment (KeithWoods) [inline src/router/Router.js]: should  blow up if this isn't a function, i.e. simple `Guard.isFunction(handler)`. 

Looks like most other public methods have similar (except `runAction` which should).

File: tests/router/router.onErrorHandlerTests.js [hunk 6]
```diff
@@ -0,0 +1,225 @@
+// notice_start
+/*
```
Comment (KeithWoods) [inline tests/router/router.onErrorHandlerTests.js]: These tests look great

**Comment by KeithWoods** [comment]
Thanks Ray.

File: src/router/Router.js (hunk 5 above)
Comment (raybooysen) [inline src/router/Router.js]: I forgot about this.  Shall I add another PR?

Comment (KeithWoods) [inline src/router/Router.js]: it's not the end of the world. The call to this handler is in a try catch so we should be ok

Comment (KeithWoods) [inline src/router/Router.js]: I might just add an issue to fix `runAction` and this at the same time

**Pull request merged**